src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -952,9 +1593,28 @@ static bool read_and_normalize_uint32 (uint32_t * __restrict val, char * __restr
   return true;
 }
 
-static bool normalize_uint64 (char * __restrict data, uint32_t * __restrict off, uint32_t size, bool bswap)
+static bool peek_and_normalize_uint32 (uint32_t * __restrict val, char * __restrict data, uint32_t * __restrict off, uint32_t size, bool bswap)
+{
+  if ((*off = check_align_prim (*off, size, 2)) == UINT32_MAX)
+    return false;
+  if (bswap)
+    *val = ddsrt_bswap4u (*((uint32_t *) (data + *off)));
+  else
+    *val = *((uint32_t *) (data + *off));
+  return true;
+}
+
+static bool normalize_enum (char * __restrict data, uint32_t * __restrict off, uint32_t size, bool bswap, uint32_t max)
+{
+  uint32_t val;
+  if (!read_and_normalize_uint32 (&val, data, off, size, bswap))
+    return false;
+  return val <= max;
+}
+
+static bool normalize_uint64 (char * __restrict data, uint32_t * __restrict off, uint32_t size, bool bswap, uint32_t xcdr_version)
 {
-  if ((*off = check_align_prim (*off, size, 3)) == UINT32_MAX)
+  if ((*off = check_align_prim (*off, size, xcdr_version == CDR_ENC_VERSION_2 ? 2 : 3)) == UINT32_MAX)
     return false;
   if (bswap)
     *((uint64_t *) (data + *off)) = ddsrt_bswap8u (*((uint64_t *) (data + *off)));
```

Comment:
This 64-bit access needs to be split, too. (I hope that I haven't overlooked any so far ... but I am not sure.)

Reply:
Fixed in 1b3eb65